<?xml version="1.0" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>batch_findmite.pl</title>
<link rel="stylesheet" href="../css/dawgpaws_man.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:_www@b04.apple.com" />
</head>

<body>
<table border="0" width="100%" cellspacing="0" cellpadding="3">
<tr><td class="block" valign="middle">
<big><strong><span class="block">&nbsp;batch_findmite.pl</span></strong></big>
</td></tr>
</table>

<p><a name="__index__"></a></p>
<!-- INDEX BEGIN -->

<ul>

	<li><a href="#version">VERSION</a></li>
	<li><a href="#synopsis">SYNOPSIS</a></li>
	<ul>

		<li><a href="#usage">Usage</a></li>
		<li><a href="#required_arguments">Required Arguments</a></li>
	</ul>

	<li><a href="#description">DESCRIPTION</a></li>
	<li><a href="#required_arguments">REQUIRED ARGUMENTS</a></li>
	<li><a href="#options">OPTIONS</a></li>
	<li><a href="#diagnostics">DIAGNOSTICS</a></li>
	<li><a href="#configuration_and_environment">CONFIGURATION AND ENVIRONMENT</a></li>
	<ul>

		<li><a href="#configuration_file">Configuration File</a></li>
	</ul>

	<li><a href="#dependencies">DEPENDENCIES</a></li>
	<ul>

		<li><a href="#required_software">Required Software</a></li>
		<li><a href="#required_perl_modules">Required Perl Modules</a></li>
	</ul>

	<li><a href="#bugs_and_limitations">BUGS AND LIMITATIONS</a></li>
	<ul>

		<li><a href="#limitations">Limitations</a></li>
	</ul>

	<li><a href="#see_also">SEE ALSO</a></li>
	<li><a href="#reference">REFERENCE</a></li>
	<li><a href="#license">LICENSE</a></li>
	<li><a href="#author">AUTHOR</a></li>
	<li><a href="#history">HISTORY</a></li>
</ul>
<!-- INDEX END -->

<hr />
<p>
</p>
<h1><a name="version">VERSION</a></h1>
<p>This documentation refers to batch_findmite version Release 1.0</p>
<p>
</p>
<hr />
<h1><a name="synopsis">SYNOPSIS</a></h1>
<p>
</p>
<h2><a name="usage">Usage</a></h2>
<pre>
    batch_findmite.pl -i InDir -o OutDir -c ConfigFile [-gff]</pre>
<p>
</p>
<h2><a name="required_arguments">Required Arguments</a></h2>
<pre>
    -i, --indir    # Directory of fasta files to process
    -o, --outdir   # Path to the base output directory
    -c, --config   # Path to the config file
    --gff          # Produce output in GFF format</pre>
<p>
</p>
<hr />
<h1><a name="description">DESCRIPTION</a></h1>
<p>The batch_findmite program will do a FINDMITE analysis for each parameter
set in your configuration file for each query sequence in your 
input directory. The results from FINDMITE have a VERY high false positive
rate so you will need to further evaluate your results to find the true MITEs
in your query sequence.</p>
<p>
</p>
<hr />
<h1><a name="required_arguments">REQUIRED ARGUMENTS</a></h1>
<dl>
<dt><strong><a name="item__2di_2c_2d_2dindir">-i,--indir</a></strong>

<dd>
<p>Path of the input directory. This is a directory that contains the fasta
files to process.</p>
</dd>
</li>
<dt><strong><a name="item__2do_2c_2d_2doutdir">-o,--outdir</a></strong>

<dd>
<p>Path of the base output directory.</p>
</dd>
</li>
<dt><strong><a name="item__2dc_2c_2d_2dconfig">-c,--config</a></strong>

<dd>
<p>Path to the configuration file that includes the parameter sets
for running the FINDMITE program.
These parameters represent the answers to the series of
questions that must be answered when running FINDMITE. Any lines
starting with # are ignored.</p>
</dd>
<dd>
<p><strong>EXAMPLE</strong></p>
</dd>
<dd>
<pre>
  #------------------------------------------------------------------
  # Name        Rep     TIR  Mis AT GC ATTA     2Base   Min     Max
  #------------------------------------------------------------------
  TA_11         TA      11   1   y  y  y        85      30      700
  TA_12         TA      12   1   y  y  y        85      30      700</pre>
</dd>
<dd>
<p>For a detail description, see the Paramters File heading under 
the CONFIGURATION AND ENVIRONMENT section of the full program 
documentation.</p>
</dd>
</li>
</dl>
<p>
</p>
<hr />
<h1><a name="options">OPTIONS</a></h1>
<dl>
<dt><strong><a name="item__2df_2c_2d_2dfasta">-f,--fasta</a></strong>

<dd>
<p>Create a fasta file of all predicted MITEs. A different fasta file will be 
created for each of the parameter set names. This will currently append to
the existing fasta data file in the output directory.</p>
</dd>
</li>
<dt><strong><a name="item__2dgff">-gff</a></strong>

<dd>
<p>Produce a GFF output file that indicates where the predicted MITEs are on
the query sequence.</p>
</dd>
</li>
<dt><strong><a name="item__2dq_2c_2d_2dquiet">-q,--quiet</a></strong>

<dd>
<p>Run the program with minimal output. Does not require user interaction.</p>
</dd>
</li>
<dt><strong><a name="item__2d_2dverbose">--verbose</a></strong>

<dd>
<p>Run the program with maximal output.</p>
</dd>
</li>
<dt><strong><a name="item__2d_2dusage">--usage</a></strong>

<dd>
<p>Short overview of how to use program from command line.</p>
</dd>
</li>
<dt><strong><a name="item__2d_2dhelp">--help</a></strong>

<dd>
<p>Show program usage with summary of options.</p>
</dd>
</li>
<dt><strong><a name="item__2d_2dversion">--version</a></strong>

<dd>
<p>Show program version.</p>
</dd>
</li>
<dt><strong><a name="item__2d_2dman">--man</a></strong>

<dd>
<p>Show the full program manual. This uses the perldoc command to print the 
POD documentation for the program.</p>
</dd>
</li>
</dl>
<p>
</p>
<hr />
<h1><a name="diagnostics">DIAGNOSTICS</a></h1>
<p>Error messages generated by this program and possible solutions are listed
below.</p>
<dl>
<dt><strong><a name="item_error_3a_no_fasta_files_were_found_in_the_input_di">ERROR: No fasta files were found in the input directory</a></strong>

<dd>
<p>The input directory does not contain fasta files in the expected format.
This could happen because you gave an incorrect path or because your sequence 
files do not have the expected *.fasta extension in the file name.</p>
</dd>
</li>
<dt><strong><a name="item_error_3a_could_not_create_the_output_directory">ERROR: Could not create the output directory</a></strong>

<dd>
<p>The output directory could not be created at the path you specified. 
This could be do to the fact that the directory that you are trying
to place your base directory in does not exist, or because you do not
have write permission to the directory you want to place your file in.</p>
</dd>
</li>
</dl>
<p>
</p>
<hr />
<h1><a name="configuration_and_environment">CONFIGURATION AND ENVIRONMENT</a></h1>
<p>
</p>
<h2><a name="configuration_file">Configuration File</a></h2>
<p>The path to the configuration file is indicated at the command line with
-c or --config.</p>
<p>This file is a space delimited text
file that indicates the parameters to use when running the findmite
program. These parameters represent the answers to the series of
questions that must be answered when running FINDMITE.</p>
<p><strong>EXAMPLE</strong></p>
<pre>
  #------------------------------------------------------------------
  # Name        Rep     TIR  Mis AT GC ATTA     2Base   Min     Max
  #------------------------------------------------------------------
  TA_11         TA      11   1   y  y  y        85      30      700
  TA_12         TA      12   1   y  y  y        85      30      700</pre>
<p>The columns above represent the following information:</p>
<dl>
<dt><strong><a name="item_col_2e_1">Col. 1</a></strong>

<dd>
<p>Base name to assign to putative mites</p>
</dd>
</li>
<dt><strong><a name="item_col_2e_2">Col. 2</a></strong>

<dd>
<p>Direct Repeat</p>
</dd>
</li>
<dt><strong><a name="item_col_2e_3">Col. 3</a></strong>

<dd>
<p>Length of the Terminal Inverted Repeat (TIR)</p>
</dd>
</li>
<dt><strong><a name="item_col_2e_4">Col. 4</a></strong>

<dd>
<p>Number of mismatches</p>
</dd>
</li>
<dt><strong><a name="item_col_2e_5">Col. 5</a></strong>

<dd>
<p>Boolean to fileter the A/T.
This must be set to y or n.</p>
</dd>
</li>
<dt><strong><a name="item_col_2e_6">Col. 6</a></strong>

<dd>
<p>Boolean to Filter C/G
This must be set to y or n.</p>
</dd>
</li>
<dt><strong><a name="item_col_2e_7">Col. 7</a></strong>

<dd>
<p>Boolean to filter AT/TA
This must be set to y or n.</p>
</dd>
</li>
<dt><strong><a name="item_col_2e_8">Col. 8</a></strong>

<dd>
<p>Proporiton of 2Base to filter.
This must be an integer between 0 and 100.</p>
</dd>
</li>
<dt><strong><a name="item_col_2e_9">Col. 9</a></strong>

<dd>
<p>Minimum distance between TIRs
This must be an integer.</p>
</dd>
</li>
<dt><strong><a name="item_col_2e_10">Col. 10</a></strong>

<dd>
<p>Maximum distance between TIRs
This must be an integer.</p>
</dd>
</li>
</dl>
<p>
</p>
<hr />
<h1><a name="dependencies">DEPENDENCIES</a></h1>
<p>
</p>
<h2><a name="required_software">Required Software</a></h2>
<ul>
<li><strong><a name="item_findmite">FINDMITE</a></strong>

<p>The batch_findmite.pl program is dependent on the FINDMITE program. 
A version of 
FINDMITE compiled for RedHat Linux is available at:
<a href="http://jaketu.biochem.vt.edu/dl_software.htm">http://jaketu.biochem.vt.edu/dl_software.htm</a></p>
</li>
</ul>
<p>
</p>
<h2><a name="required_perl_modules">Required Perl Modules</a></h2>
<ul>
<li><strong><a name="item_file_3a_3acopy">File::Copy</a></strong>

<p>This module is required to copy the BLAST results.</p>
</li>
<li><strong><a name="item_getopt_3a_3along">Getopt::Long</a></strong>

<p>This module is required to accept options at the command line.</p>
</li>
</ul>
<p>
</p>
<hr />
<h1><a name="bugs_and_limitations">BUGS AND LIMITATIONS</a></h1>
<ul>
<li><strong><a name="item_no_bugs_currently_known">No bugs currently known</a></strong>

<p>If you find a bug with this software, file a bug report on the DAWG-PAWS
Sourceforge website: <a href="http://sourceforge.net/tracker/?group_id=204962">http://sourceforge.net/tracker/?group_id=204962</a></p>
</li>
</ul>
<p>
</p>
<h2><a name="limitations">Limitations</a></h2>
<ul>
<li><strong><a name="item_config_file_must_use_unix_format_line_endings">Config file must use UNIX format line endings</a></strong>

<p>The config file must have UNIX formatted line endings. Because of
this any config files that have been edited in programs such as
MS Word must be converted to a UNIX compatible text format before
being used with batch_blast.</p>
</li>
</ul>
<p>
</p>
<hr />
<h1><a name="see_also">SEE ALSO</a></h1>
<p>The batch_blast.pl program is part of the DAWG-PAWS package of genome
annotation programs. See the DAWG-PAWS web page 
( <a href="http://dawgpaws.sourceforge.net/">http://dawgpaws.sourceforge.net/</a> )
or the Sourceforge project page 
( <a href="http://sourceforge.net/projects/dawgpaws">http://sourceforge.net/projects/dawgpaws</a> ) 
for additional information about this package.</p>
<p>
</p>
<hr />
<h1><a name="reference">REFERENCE</a></h1>
<p>A manuscript is being submitted describing the DAWGPAWS program. 
Until this manuscript is published, please refer to the DAWGPAWS 
SourceForge website when describing your use of this program:</p>
<p>JC Estill and JL Bennetzen. 2009. 
The DAWGPAWS Pipeline for the Annotation of Genes and Transposable 
Elements in Plant Genomes.
<a href="http://dawgpaws.sourceforge.net/">http://dawgpaws.sourceforge.net/</a></p>
<p>
</p>
<hr />
<h1><a name="license">LICENSE</a></h1>
<p>GNU GENERAL PUBLIC LICENSE, VERSION 3</p>
<p><a href="http://www.gnu.org/licenses/gpl.html">http://www.gnu.org/licenses/gpl.html</a></p>
<p>THIS SOFTWARE COMES AS IS, WITHOUT ANY EXPRESS OR IMPLIED
WARRANTY. USE AT YOUR OWN RISK.</p>
<p>
</p>
<hr />
<h1><a name="author">AUTHOR</a></h1>
<p>James C. Estill &lt;JamesEstill at gmail.com&gt;

</p>
<p>
</p>
<hr />
<h1><a name="history">HISTORY</a></h1>
<p>STARTED: 08/30/2007

</p>
<p>UPDATED: 03/24/2009

</p>
<p>VERSION: Release 1.0

</p>
<table border="0" width="100%" cellspacing="0" cellpadding="3">
<tr><td class="block" valign="middle">
<big><strong><span class="block">&nbsp;batch_findmite.pl</span></strong></big>
</td></tr>
</table>

</body>

</html>
