<?xml version="1.0" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>batch_hmmer.pl</title>
<link rel="stylesheet" href="../css/dawgpaws_man.css" type="text/css" />
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:_www@b04.apple.com" />
</head>

<body>
<table border="0" width="100%" cellspacing="0" cellpadding="3">
<tr><td class="block" valign="middle">
<big><strong><span class="block">&nbsp;batch_hmmer.pl</span></strong></big>
</td></tr>
</table>

<p><a name="__index__"></a></p>
<!-- INDEX BEGIN -->

<ul>

	<li><a href="#name">NAME</a></li>
	<li><a href="#version">VERSION</a></li>
	<li><a href="#synopsis">SYNOPSIS</a></li>
	<ul>

		<li><a href="#usage">Usage</a></li>
		<li><a href="#required_arguments">Required Arguments</a></li>
	</ul>

	<li><a href="#description">DESCRIPTION</a></li>
	<li><a href="#required_arguments">REQUIRED ARGUMENTS</a></li>
	<li><a href="#options">OPTIONS</a></li>
	<li><a href="#diagnostics">DIAGNOSTICS</a></li>
	<li><a href="#configuration_and_environment">CONFIGURATION AND ENVIRONMENT</a></li>
	<ul>

		<li><a href="#configuration_file">Configuration File</a></li>
	</ul>

	<li><a href="#dependencies">DEPENDENCIES</a></li>
	<ul>

		<li><a href="#required_software">Required Software</a></li>
		<li><a href="#required_perl_modules">Required Perl Modules</a></li>
	</ul>

	<li><a href="#bugs_and_limitations">BUGS AND LIMITATIONS</a></li>
	<ul>

		<li><a href="#bugs">Bugs</a></li>
		<li><a href="#limitations">Limitations</a></li>
	</ul>

	<li><a href="#see_also">SEE ALSO</a></li>
	<li><a href="#reference">REFERENCE</a></li>
	<li><a href="#license">LICENSE</a></li>
	<li><a href="#author">AUTHOR</a></li>
	<li><a href="#history">HISTORY</a></li>
</ul>
<!-- INDEX END -->

<hr />
<p>
</p>
<h1><a name="name">NAME</a></h1>
<p>batch_hmmer.pl - Run hmmsearch in batch mode</p>
<p>
</p>
<hr />
<h1><a name="version">VERSION</a></h1>
<p>This documentation refers to program version Release 1.0</p>
<p>
</p>
<hr />
<h1><a name="synopsis">SYNOPSIS</a></h1>
<p>
</p>
<h2><a name="usage">Usage</a></h2>
<pre>
    batch_hmmer.pl -i InDir -o OutDir -c Config.cfg [--gff]</pre>
<p>
</p>
<h2><a name="required_arguments">Required Arguments</a></h2>
<pre>
    --indir         # Path to the input dir containing fasta files
    --outdir        # Path to the base output directory
    --config        # Configuration file for running hmmer
    --gff           # Path to gff output file</pre>
<p>
</p>
<hr />
<h1><a name="description">DESCRIPTION</a></h1>
<p>Given a config file this will run the hmmsearch program for each parameter
set in the configuration file for each fasta sequence file in the input
directory. This will also convert the results to GFF format if requested.</p>
<p>
</p>
<hr />
<h1><a name="required_arguments">REQUIRED ARGUMENTS</a></h1>
<dl>
<dt><strong><a name="item__2di_2c_2d_2dindir">-i,--indir</a></strong>

<dd>
<p>Path of the input directory. This is the directory that contains
the</p>
</dd>
</li>
<dt><strong><a name="item__2do_2c_2d_2doutdir">-o,--outdir</a></strong>

<dd>
<p>Path of the base output directory.</p>
</dd>
</li>
<dt><strong><a name="item__2dc_2c_2d_2dconfig">-c,--config</a></strong>

<dd>
<p>Path of the configuration file specifying the parameters used to run
the hmmer program. This config file includes (1) the name of the
parameter set, (2) the directory of the hmm models, and (3)
command line arguments for hmmsearch.</p>
</dd>
</li>
</dl>
<p>
</p>
<hr />
<h1><a name="options">OPTIONS</a></h1>
<dl>
<dt><strong><a name="item__2d_2dgff">--gff</a></strong>

<dd>
<p>Produce a GFF output file for each of the query sequence for each
of the parameter sets in the input file.</p>
</dd>
</li>
<dt><strong><a name="item__2dq_2c_2d_2dquiet">-q,--quiet</a></strong>

<dd>
<p>Run the program with minimal output.</p>
</dd>
</li>
<dt><strong><a name="item__2d_2dverbose">--verbose</a></strong>

<dd>
<p>Run the program in verbose mode.</p>
</dd>
</li>
<dt><strong><a name="item__2d_2dversion">--version</a></strong>

<dd>
<p>Show program version.</p>
</dd>
</li>
<dt><strong><a name="item__2d_2dusage">--usage</a></strong>

<dd>
<p>Short overview of how to use program from command line.</p>
</dd>
</li>
<dt><strong><a name="item__2d_2dhelp">--help</a></strong>

<dd>
<p>Show program usage with summary of options.</p>
</dd>
</li>
<dt><strong><a name="item__2d_2dman">--man</a></strong>

<dd>
<p>Show the full program manual. This uses the perldoc command to print the 
POD documentation for the program.</p>
</dd>
</li>
</dl>
<p>
</p>
<hr />
<h1><a name="diagnostics">DIAGNOSTICS</a></h1>
<p>Error messages generated by this program and possible solutions are listed
below.</p>
<dl>
<dt><strong><a name="item_error_3a_no_fasta_files_were_found_in_the_input_di">ERROR: No fasta files were found in the input directory</a></strong>

<dd>
<p>The input directory does not contain fasta files in the expected format.
This could happen because you gave an incorrect path or because your sequence 
files do not have the expected *.fasta extension in the file name.</p>
</dd>
</li>
<dt><strong><a name="item_error_3a_could_not_create_the_output_directory">ERROR: Could not create the output directory</a></strong>

<dd>
<p>The output directory could not be created at the path you specified. 
This could be do to the fact that the directory that you are trying
to place your base directory in does not exist, or because you do not
have write permission to the directory you want to place your file in.</p>
</dd>
</li>
</dl>
<p>
</p>
<hr />
<h1><a name="configuration_and_environment">CONFIGURATION AND ENVIRONMENT</a></h1>
<p>
</p>
<h2><a name="configuration_file">Configuration File</a></h2>
<p>The location of the configuration file is indicated by the --config
option at the command line. This is a tab delimited text file. 
The columns of data represent  the following information</p>
<dl>
<dt><strong><a name="item_col_1_2e">col 1.</a></strong>

<dd>
<p>Name of the parameter set</p>
</dd>
</li>
<dt><strong><a name="item_col_2_2e_dir_of_the_hmmer_models">col 2. Dir of the hmmer models</a></strong>

<dd>
<p>All hmm profile models in this directory will be searched using the
hmmer arguments in col 3.</p>
</dd>
</li>
<dt><strong><a name="item_col_3_2earguments_for_hmmer">col 3.Arguments for hmmer</a></strong>

<dd>
<p>The arguments for hmmer should be separate by spaces and can include the 
following arguments:</p>
</dd>
<dl>
<dt><strong><a name="item__2da__3cn_3e">-A &lt;n&gt;</a></strong>

<dd>
<p>sets alignment output limit to &lt;n&gt; best domain alignments</p>
</dd>
</li>
<dt><strong><a name="item__2de__3cx_3e">-E &lt;x&gt;</a></strong>

<dd>
<p>sets E value cutoff (globE) to &lt;= x</p>
</dd>
</li>
<dt><strong><a name="item__2dt__3cx_3e">-T &lt;x&gt;</a></strong>

<dd>
<p>sets T bit threshold (globT) to &gt;= x</p>
</dd>
</li>
<dt><strong><a name="item__2dz__3cn_3e">-Z &lt;n&gt;</a></strong>

<dd>
<p>sets Z (# seqs) for E-value calculation</p>
</dd>
</li>
<dt><strong><a name="item__2d_2dcompat">--compat</a></strong>

<dd>
<p>make best effort to use last version's output style</p>
</dd>
</li>
<dt><strong><a name="item__2d_2dcpu__3cn_3e">--cpu &lt;n&gt;</a></strong>

<dd>
<p>run &lt;n&gt; threads in parallel (if threaded)</p>
</dd>
</li>
<dt><strong><a name="item__2d_2dcut_ga">--cut_ga</a></strong>

<dd>
<p>use Pfam GA gathering threshold cutoffs</p>
</dd>
</li>
<dt><strong><a name="item__2d_2dcut_nc">--cut_nc</a></strong>

<dd>
<p>use Pfam NC noise threshold cutoffs</p>
</dd>
</li>
<dt><strong><a name="item__2d_2dcut_tc">--cut_tc</a></strong>

<dd>
<p>use Pfam TC trusted threshold cutoffs</p>
</dd>
</li>
<dt><strong><a name="item__2d_2ddome__3cx_3e">--domE &lt;x&gt;</a></strong>

<dd>
<p>sets domain Eval cutoff (2nd threshold) to &lt;= x</p>
</dd>
</li>
<dt><strong><a name="item__2d_2ddomt__3cx_3e">--domT &lt;x&gt;</a></strong>

<dd>
<p>sets domain T bit thresh (2nd threshold) to &gt;= x</p>
</dd>
</li>
<dt><strong><a name="item__2d_2dforward">--forward</a></strong>

<dd>
<p>use the full <code>Forward()</code> algorithm instead of Viterbi</p>
</dd>
</li>
<dt><strong><a name="item__2d_2dinformat__3cs_3e">--informat &lt;s&gt;</a></strong>

<dd>
<p>sequence file is in format &lt;s&gt;</p>
</dd>
</li>
<dt><strong><a name="item__2d_2dnull2">--null2</a></strong>

<dd>
<p>turn OFF the post hoc second null model</p>
</dd>
</li>
<dt><strong><a name="item__2d_2dpvm">--pvm</a></strong>

<dd>
<p>run on a Parallel Virtual Machine (PVM)</p>
</dd>
</li>
<dt><strong><a name="item__2d_2dxnu">--xnu</a></strong>

<dd>
<p>turn ON XNU filtering of target protein sequences</p>
</dd>
</li>
</dl>
</dl>
<p>An example that will do hmmsearch for four directories of hmm models
using an evalue threshold of 0.00001 follows:</p>
<pre>
    rice_mite  /$HOME/HMMData/db/hmm/rice_mite_models/  -E 0.00001
    rice_mule  /$HOME/HMMData/db/hmm/rice_mule_models/  -E 0.00001
    tpase      /$HOME/HMMData/db/hmm/tpase_models/      -E 0.00001
    pfam       /$HOME/HMMData/db/hmm/pfam/              -E 0.00001</pre>
<p>
</p>
<hr />
<h1><a name="dependencies">DEPENDENCIES</a></h1>
<p>
</p>
<h2><a name="required_software">Required Software</a></h2>
<ul>
<li><strong><a name="item_hmmer">HMMER</a></strong>

<p>The hmmer program is required: <a href="http://hmmer.janelia.org/.">http://hmmer.janelia.org/.</a> Specifically
this program requires the hmmsearch program.</p>
</li>
</ul>
<p>
</p>
<h2><a name="required_perl_modules">Required Perl Modules</a></h2>
<ul>
<li><strong><a name="item_file_3a_3acopy">File::Copy</a></strong>

<p>This module is required to copy the BLAST results.</p>
</li>
<li><strong><a name="item_getopt_3a_3along">Getopt::Long</a></strong>

<p>This module is required to accept options at the command line.</p>
</li>
<li><strong><a name="item_text_3a_3awrap">Text::Wrap</a></strong>

<p>This module allows for word wrapping and hanging indents of printed
text. This allows for a more readable output for long strings</p>
</li>
<li><strong><a name="item_bio_3a_3atools_3a_3ahmmer_3a_3aresults">Bio::Tools::HMMER::Results</a></strong>

<p>This module is part of the BioPerl package. This allows for
parsing of the the output from the hmmsearch program.</p>
</li>
</ul>
<p>
</p>
<hr />
<h1><a name="bugs_and_limitations">BUGS AND LIMITATIONS</a></h1>
<p>
</p>
<h2><a name="bugs">Bugs</a></h2>
<ul>
<li><strong><a name="item_no_bugs_currently_known">No bugs currently known</a></strong>

<p>If you find a bug with this software, file a bug report on the DAWG-PAWS
Sourceforge website: <a href="http://sourceforge.net/tracker/?group_id=204962">http://sourceforge.net/tracker/?group_id=204962</a></p>
</li>
</ul>
<p>
</p>
<h2><a name="limitations">Limitations</a></h2>
<ul>
<li><strong><a name="item_must_created_profile_hmms_external">Must created profile hmms external</a></strong>

<p>The DAWG-PAWS package currently does not include programs for 
creating hmm profile models. These must be created externally using
the hmmer package of programs.</p>
</li>
<li><strong><a name="item_config_file_must_use_unix_format_line_endings">Config file must use UNIX format line endings</a></strong>

<p>The config file must have UNIX formatted line endings. Because of
this any config files that have been edited in programs such as
MS Word must be converted to a UNIX compatible text format before
being used with batch_blast.</p>
</li>
</ul>
<p>
</p>
<hr />
<h1><a name="see_also">SEE ALSO</a></h1>
<p>The batch_blast.pl program is part of the DAWG-PAWS package of genome
annotation programs. See the DAWG-PAWS web page 
( <a href="http://dawgpaws.sourceforge.net/">http://dawgpaws.sourceforge.net/</a> )
or the Sourceforge project page 
( <a href="http://sourceforge.net/projects/dawgpaws">http://sourceforge.net/projects/dawgpaws</a> ) 
for additional information about this package.</p>
<p>
</p>
<hr />
<h1><a name="reference">REFERENCE</a></h1>
<p>A manuscript is being submitted describing the DAWGPAWS program. 
Until this manuscript is published, please refer to the DAWGPAWS 
SourceForge website when describing your use of this program:</p>
<p>JC Estill and JL Bennetzen. 2009. 
The DAWGPAWS Pipeline for the Annotation of Genes and Transposable 
Elements in Plant Genomes.
<a href="http://dawgpaws.sourceforge.net/">http://dawgpaws.sourceforge.net/</a></p>
<p>
</p>
<hr />
<h1><a name="license">LICENSE</a></h1>
<p>GNU GENERAL PUBLIC LICENSE, VERSION 3</p>
<p><a href="http://www.gnu.org/licenses/gpl.html">http://www.gnu.org/licenses/gpl.html</a></p>
<p>THIS SOFTWARE COMES AS IS, WITHOUT ANY EXPRESS OR IMPLIED
WARRANTY. USE AT YOUR OWN RISK.</p>
<p>
</p>
<hr />
<h1><a name="author">AUTHOR</a></h1>
<p>James C. Estill &lt;JamesEstill at gmail.com&gt;

</p>
<p>
</p>
<hr />
<h1><a name="history">HISTORY</a></h1>
<p>STARTED: 09/17/2007

</p>
<p>UPDATED: 03/24/2009

</p>
<p>VERSION: Release 1.0

</p>
<table border="0" width="100%" cellspacing="0" cellpadding="3">
<tr><td class="block" valign="middle">
<big><strong><span class="block">&nbsp;batch_hmmer.pl</span></strong></big>
</td></tr>
</table>

</body>

</html>
